### PR TITLE
Query: Refactoring Part 1 with query filtering.

### PR DESCRIPTION
**Pull request by rmclaren**
## Description

This pull request primarily deals with 2 things:

1. Refactoring the code from the Query Parsing function (`QueryParser::parse`) up to (and including) the code that identifies the targets in the BUFR data (`QueryRunner::findTargets`). Some significant changes are listed bellow:

- `Tokenizer` class was added. It is responsible for taking a raw string the user provided and parsing it into a list of `Token` objects.
- `QueryParser` class is rewritten to use the tokenizer. It parses the `Token` objects into a list of `QueryComponent` objects and stores the result into a `Query` object.
- `SubsetTable` class was rewritten. It now re-interprets the BUFR meta data into into a recursive tree where each node represents a data element. This tree can then be interrogated to get the information we need to do our job.
- 'QueryRunner::findTargets' method was rewritten. It now uses `SubsetTable` to do its work (previously had implemented a similar, but slightly different process). `Targets` now stores `TargetComponents` which map to the different components in the query.
- Because of the new code above there is a fix/improvement in some query strings. Some BUFR tables list repeated mnemonics multiple times (ex: satwind data often lists the QCPRMS mnemonic multiple times...). Before the index for the repeat was listed at the end of the query (ex: `*/QCPRMS/GNAP[4]`). This was kinda incorrect and misleading giving the impression that GNAP was the thing that is repeated in the BUFR table. It also meant that it couldn't handle situations where there were multiple places with repeating mnemonics in a query path. The legacy code just couldn't deal with situations like this... The updated syntax now places the index were it belongs (ex: `*/QCPRMS[4]/GNAP`). 

_Note: Refactoring Part 2 will cover code from `QueryParser::collectData` onward._

2. Adds functionality which allows for the inclusion of a filter in the query string for cases when you want to select a subset of data in a particular repeat in the data (functionality dependent on refactoring). Here are some examples of what the new query strings can look like:

- `*/BRIT{2}/TMBR` Uses only the second repeat of BRIT and drops the extra dimension associated with the BRIT repeat (you get 1d not 2d dimensional data).
- `*/BRIT{1-3, 5}/TMBR` Keeps repeats 1,2,3 and 5 from BRIT. The data dimensionality does not change because you are filtering for more than 1 repeat.
- `*/ROSEQ1{3}/ROSEQ3{2-4}/PCCF` Pick the third instance of ROSEQ1 and drop that dimension, 2 thru 4 of ROSEQ3 retaining that dimension. 
- `*/QCPRMS[4]{2}/GNAP` The mnemonic for QCPRMS is repeated 4 times. Pick the BUFR data that corresponds to the 
4th instance of that mnemonic and only keep the 2nd repeat dropping that dimension.

_Note: The code that does the actual filtering can be found in `QueryParser::collectData`. Yes, it will get caught up in Refactoring Part 2._

### Issue(s) addressed

#1180 
#967 

## Acceptance Criteria (Definition of Done)

Unit tests pass. Queries like `*/ABC{1-2}/XYZ` work.

### CODE REVIEW

**Comment by rmclaren** [comment]
@CoryMartin-NOAA Would it be possible to grant Edward Hartnett access to this JCSDA internal? It might be time time make him aware of what we are doing...

**Comment by CoryMartin-NOAA** [comment]
@rmclaren I don't have the power to do that, this is a public repo though (despite it being in internal). Does he need write access or will an "@" for an FYI suffice? (if the former, we have to go through Daryl)

**Comment by rmclaren** [comment]
@CoryMartin-NOAA I don't think he needs write access, I just wanted him to take a look... I was not aware that this repo was publicly visible. Thanks:)

**Comment by rmclaren** [comment]
Note to the reviewers: Looking at the diffs for this PR is not much use as in most cases things have been completely replaced. It would probably be easier just to look at the new code directly without the diffs.

File: src/bufr/BufrParser/Query/ResultSet.cpp
Comment (nicholasesposito) [inline src/bufr/BufrParser/Query/ResultSet.cpp:286]: Why the change to > 0 instead of > 1?

File: src/bufr/BufrParser/Query/SubsetTable.h
Comment (nicholasesposito) [inline src/bufr/BufrParser/Query/SubsetTable.h:79]: What is an example of when this would happen and/or be needed?

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (nicholasesposito) [inline src/bufr/BufrParser/Query/Tokenizer.h:43]: So would a subtoken be when we're looking at something like `*/BRIT{2}/TMBR` where TMBR  is the subtoken?

**Review by nicholasesposito — COMMENTED** [review]
Reviewing and I have some questions. I'll likely have more as I continue to dig. Looks good so far though.

**Comment by BenjaminRuston** [comment]
this is a lot of files changed at once, any way to break it down further or the remaining parts?

File: doc/bufr_uml/BigPicture.puml
Comment (rmclaren) [reply doc/bufr_uml/BigPicture.puml:1]: These are plantuml files. PlantUml is a library which lets you specify UML images via text files. They can be rendered using the plantuml executable. They are basically like source files. I'll leave it up to the group wether we want to check in the resulting images as well...

File: src/bufr/BufrParser/Query/ResultSet.cpp
Comment (rmclaren) [reply src/bufr/BufrParser/Query/ResultSet.cpp:286]: I was accidentally not slicing if there is only 1 dataframe in the file.

Comment (rmclaren) [reply src/bufr/BufrParser/Query/ResultSet.cpp:286]: By slicing I mean removing unwanted dimensional data from the dimensions array.

File: src/bufr/BufrParser/Query/SubsetTable.h
Comment (rmclaren) [reply src/bufr/BufrParser/Query/SubsetTable.h:79]: This function tells you if a node is a type of node that can contain children. It can be useful in a lot of ways, for example while recursively searching the BUFR subset tree you can use it to figure out if you need to dive deeper or not.

File: src/bufr/BufrParser/Query/SubsetTable.h
Comment (rmclaren) [reply src/bufr/BufrParser/Query/SubsetTable.h:131]: Parent is a weak_ptr. This call checks to make sure the weak_ptr is set to a valid pointer. In practice they will be expired if you are at the root of the tree.

File: doc/bufr_uml/BigPicture.puml
Comment (emilyhcliu) [reply doc/bufr_uml/BigPicture.puml:1]: @rmclaren Same question as @nicholasesposito It would be great if we could visualize the plantuml files.  
The diagrams may help us understand the new code.

**Comment by emilyhcliu** [comment]
**Big Picture**

VPD1RnCn48Nlyoj6FQM1L9MuLT6gWhOI1rGG5VV9UfAQUcrhx4WKjFyTkydMim7raCE-xsNcUSQ-JfaaxmS7ZzOmyX3pCmwS7uD9a9wiZoGqe8aMFrnU0eJDJ-upVY7h0Q1tb1BUKARx5ttzqC4VG8HHiq_1v-9dcPWFcT7nDb_XlUp8sz-Sq1GdMfy3zfHv5yHo0e

**Big Picture Sequence**

TP7FJiCm3CRlVWeVuL2f_3cXgXeSa3XmGDiBXDOIqTAqnAvWbBqx6XdEieYZVz_lYsHdnMA2p8C3tbe_cM06x2UBDtKDt1b7U5LVt-A7xUKDe7E66JVZbhpzeW0B8EB2OiHs-yWPvHXVlCyKTfUmptgmd847mzHisbF-SYWMzlFiv9_4qp1wswKWyYVVqsTXhscYO6GCHNBlT-lvbKcA

**Data Collection**

fLLDRzim3BtxL_1Kk9i3T1-d8JFQhGZG0IlQhjkrK2mc5IfBhaWtoORzzqcM4zcjTzbEYTvxV9H8IZub5exRMcVqe4mZdAX1DWhUdPnaL0cDyERaxNlOACdtMTQ8wa6i4QvRT3kOpnTMItHb2Tcl3A3IWWWMIgDV103WDMwnQXdpoXhY6FSD6OuU-ty5aFg9Tmm6jtmC7-06gTNicHZl

**Query Class Diagram**

dLHHRziW67mslmAbKfLEdLPjRtkOccIAD6dQkZRx0GG-9oW4N0nT-k_tWHt0TH9DUsdm-Uw-uy2zRnsppkyLsKe1j3R6hSs1j3kf6sRPdef6qaztTmNo1CyUD8SPTyREZHLWLrl9TnhQbdwyJ5mOPIozggM5jKLDmUvOZEysrdWjUgQIcwrx1QNCduBxW9abbmtJxZyKRoDVeeq24nPo

**Get Result Set** 

dL9DJ-Cm4BtxLuoqb_QmgsgvvO3w0K98L2eg5uwkFKcifdQOZ87-UvmqQcmQUk2McVSnJo-UDoo9_Q4IfT48kNEySv-YsHjRIv87qBM1czaiscpnpQDLk53iQEb88xsKHkqjDWtylqvSkSeH_CaDuOw29cAVM5Bj2tBUwfvPcQBa8rQL-uYu6qbibAcbvLyehaIEQUE1XMnpmbeQA-g1

**Tokenizer**

bLDDQzj043rxVyL08CJWG6XlFPJ43OPUsf2KtiVQaRtrQaUTdOtjVz-LFoAvKWGvIJozZxSFxNrK54sLDsjd2KfcNVBEn8qBDGfMO6i7d-_kEiWB_KqK2deeb6NEOab-hLsn2HGZV1edVcFF0bUb4rfArdJOHnOMcvLm2lR4z6wrrZrvpziEzmb5NU5g3FfnHQ_o2ArVs63J4-Qfb4oE

**Comment by emilyhcliu** [comment]
@rmclaren Are you addressing [Issue #967](https://github.com/JCSDA-internal/ioda-converters/issues/967) in this PR?
If I understood correctly, this PR does not address Issue #967.

**Comment by BenjaminRuston** [comment]
@CoryMartin-NOAA to add Edward Hartnett to internal you need to have Daryl make a change request

**Comment by PatNichols** [comment]
@rmclaren Don't worry about the failing tests. It's on our end.

**Comment by srherbener** [comment]
> @rmclaren sorry about this but I won't be able to review this week because of my priority being the upcoming skylab release. I'm targeting to get the review done by the middle of next week. If that is too much of a delay, please feel free to assign another reviewer.

I was able to start looking at this but didn't make much progress yet. I'll continue next week. Sorry about the delay.

**Comment by rmclaren** [comment]
@srherbener No problem! Please don't feel like you need to rush.

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (rmclaren) [reply src/bufr/BufrParser/Query/Tokenizer.h:43]: So this returns a list of QueryToken objects. So for example if you've got a MultiQuery like `[*/BRITCSTC/TMBR, */BRIT/TMBR]` then the query tokens are `*/BRITCSTC/TMBR` and `*/BRIT/TMBR`.

**Comment by rmclaren** [comment]
I updated some of the plantUML files so that they are more up to data.

**Create a new QuerySet**
CreateQuerySet

**Collecting Data**
DataCollection

**Class Diagram**
QueryClassDiagram

**Comment by rmclaren** [comment]
@BenjaminRuston Sorry for all the code changes. My ultimate goal for this was to get the filtering working. In order to get there it made sense to refactor a lot of code that I've been meaning to improve. Much like pulling a thread in a knitted sweater an improvement to one class or module will demand changes in the next one in the chain...

It would be a bit of work to try separate things out into smaller pull requests at this point, as I'd have to write compatibility layers (new to old code) at each step of the way. I'll try to break the next round of changes into smaller chunks, however.

Like I said previously, you can pretty much take the refactored code and treat it like new code... If you want to make it easier on yourself you could just focus one class at a time in the order they appear within the sequence diagrams. So for example you could start with "Create a new QuerySet" (above) and work through those classes one at a time... And then you could go through the Data Collection sequence and work through those classes.

**Comment by PatNichols** [comment]
@rmclaren We finally got the CI testing problem fixed and your PR passesd. We used it for a test actually :)

**Comment by rmclaren** [comment]
@PatNichols Cool! Thanks @PatNichols.

**Comment by emilyhcliu** [comment]
I checked out the branch and ran a test with a few satellite radiances data (GMI, ATMS, ...etc). 
I do not see any problem.

File: src/bufr/BufrParser/Query/QueryRunner.cpp
Comment (CoryMartin-NOAA) [inline src/bufr/BufrParser/Query/QueryRunner.cpp:101]:
```suggestion
                // Print message to inform the user of the missing targets
```

File: src/bufr/BufrParser/Query/Target.h
Comment (CoryMartin-NOAA) [inline src/bufr/BufrParser/Query/Target.h:45]: is `TYP` type? if so, why not just write out `type` in the comments?

File: src/bufr/BufrParser/Query/Tokenizer.cpp
Comment (CoryMartin-NOAA) [inline src/bufr/BufrParser/Query/Tokenizer.cpp:2]: 2023

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (CoryMartin-NOAA) [inline src/bufr/BufrParser/Query/Tokenizer.h:2]: 2023

File: test/testinput/bufr_ncep_adpsfc.yaml
Comment (CoryMartin-NOAA) [inline test/testinput/bufr_ncep_adpsfc.yaml:15]: why the double comment?

File: test/testinput/bufr_query_filtering.yaml
Comment (CoryMartin-NOAA) [inline test/testinput/bufr_query_filtering.yaml:1]: 2023

**Review by CoryMartin-NOAA — COMMENTED** [review]
Code looks good, have a few minor squabbles on some comments and one question, but otherwise I think I am ready to approve once these superficial fixes are made

File: doc/bufr_uml/BigPicture.puml
Comment (rmclaren) [reply doc/bufr_uml/BigPicture.puml:1]: @emilyhcliu Checked in the PNG files that go along with the plantUML files.

**Review by srherbener — CHANGES_REQUESTED** [review]
Thanks for the refactoring and adding the ctests! I think in general the code looks good but I'm seeing test failures on my Mac. Here's a listing from the `test_iodaconv_bufr_ncep_atms` test:

```
26242 665/850 Testing: test_iodaconv_bufr_ncep_atms
26243 665/850 Test: test_iodaconv_bufr_ncep_atms
26244 Command: "/usr/local/bin/bash" "/Users/steveherbener/projects/CONVERTERS/ioda-bundle/build/bin/iodaconv_comp.sh" "netcdf" "/Users/steveherbener/projects/CONVERTERS/ioda-bundle/build/bin/bufr2ioda.x testinput/bufr_ncep_atms.yaml" "gdas.t00z.atms_n20.tm00.nc" "0.0"
26245 Directory: /Users/steveherbener/projects/CONVERTERS/ioda-bundle/build/iodaconv/test
26246 "test_iodaconv_bufr_ncep_atms" start time: Apr 14 10:58 MDT
26247 Output:
26248 ----------------------------------------------------------
26249 PRE-MAIN-INFO BufrParser: Parsing file ./testinput/gdas.t00z.atms.tm00.bufr_d
26250 libc++abi: terminating due to uncaught exception of type std::__1::regex_error: One of *?+{ was not preceded by a valid regular expression.
26251 /Users/steveherbener/projects/CONVERTERS/ioda-bundle/build/bin/iodaconv_comp.sh: line 20:  2144 Abort trap: 6           $cmd
26252 <end of output>
26253 Test time =   0.34 sec
26254 ----------------------------------------------------------
26255 Test Failed.
26256 "test_iodaconv_bufr_ncep_atms" end time: Apr 14 10:58 MDT
26257 "test_iodaconv_bufr_ncep_atms" time elapsed: 00:00:00
26258 ----------------------------------------------------------
```

I'm seeing these warnings from the apple clang compilers (not sure if they are relevant):
```
/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/ResultSet.cpp:266:17: warning: logical not is only applied to the left hand side of this comparison [-Wlogical-not-parentheses]
            if (!targetField.data.size() == 0) {
                ^                        ~~
/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/ResultSet.cpp:266:17: note: add parentheses after the '!' to evaluate the comparison first
            if (!targetField.data.size() == 0) {
                ^
                 (                           )
/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/ResultSet.cpp:266:17: note: add parentheses around left hand side expression to silence this warning
            if (!targetField.data.size() == 0) {
                ^
                (                       )



/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/SubsetTable.h:22:5: warning: class 'PathComponent' was previously declared as a struct; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
    class PathComponent;
    ^
/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/QueryParser.h:72:12: note: previous use is here
    struct PathComponent : public QueryComponent
           ^
/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/SubsetTable.h:22:5: note: did you mean struct here?
    class PathComponent;
    ^~~~~
    struct



/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider/../QuerySet.h:20:5: warning: class 'Query' was previously declared as a struct; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
    class Query;
    ^
/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/QueryParser.h:119:12: note: previous use is here
    struct Query
           ^
/Users/steveherbener/projects/CONVERTERS/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider/../QuerySet.h:20:5: note: did you mean struct here?
    class Query;
    ^~~~~
    struct
```
I'm running on an M1 Mac using the Rosetta intel (x86_64) emulator. I have Ventura 13.3 installed with command line tools version 14.1 rc2 which has Apple Clang version 14.0.0. I'm also using GNU Fortran 12.2.0. I'm using version 1.3.0 of spack-stack.

**Comment by srherbener** [comment]
Everyone I am on leave for the next two weeks (return on Mon, May 1st). I am fine approving once it is working on the Mac. Please feel free to dismiss my "request changes" once the tests pass on the Mac.

**Review by emilyhcliu — APPROVED** [review]

**Comment by rmclaren** [comment]
Is there some trick to building oops with the llvm clang compiler, I have all kinds of problems... For example I keep getting the following when building the oops tests `ld: file not found: @rpath/libquadmath.0.dylib for architecture x86_64` even when I've confirmed, via otools, that the `liboops.dylib` file rpath has a valid path for libquadmath.

When I comment out the tests in the ./src/CMakeLists.txt file I then get a little further but end up with this:

```
[ 48%] Linking CXX shared library ../../lib/libqg.dylib
ld: warning: ld: warning: could not create compact unwind for ___qg_fields_mod_MOD_qg_fields_read_file: stack subq instruction is too different from dwarf stack size
could not create compact unwind for ___qg_error_covariance_mod_MOD_qg_error_covariance_setup: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for ___qg_fields_mod_MOD_qg_fields_write_file: stack subq instruction is too different from dwarf stack size
ld: warning: could not create compact unwind for ___qg_obsdb_mod_MOD_qg_obsdb_setup: stack subq instruction is too different from dwarf stack size
```

Its using the gcc version 12 version of gfortran. Weird thing all this stuff works correctly when I just use GCC/G++ tools. So I'm a bit stuck...

**Comment by climbfuji** [comment]
> Is there some trick to building oops with the llvm clang compiler, I have all kinds of problems... For example I keep getting the following when building the oops tests `ld: file not found: @rpath/libquadmath.0.dylib for architecture x86_64` even when I've confirmed, via otools, that the `liboops.dylib` file rpath has a valid path for libquadmath.
> 
> When I comment out the tests in the ./src/CMakeLists.txt file I then get a little further but end up with this:
> 
> ```
> [ 48%] Linking CXX shared library ../../lib/libqg.dylib
> ld: warning: ld: warning: could not create compact unwind for ___qg_fields_mod_MOD_qg_fields_read_file: stack subq instruction is too different from dwarf stack size
> could not create compact unwind for ___qg_error_covariance_mod_MOD_qg_error_covariance_setup: stack subq instruction is too different from dwarf stack size
> ld: warning: could not create compact unwind for ___qg_fields_mod_MOD_qg_fields_write_file: stack subq instruction is too different from dwarf stack size
> ld: warning: could not create compact unwind for ___qg_obsdb_mod_MOD_qg_obsdb_setup: stack subq instruction is too different from dwarf stack size
> ```
> 
> Its using the gcc version 12 version of gfortran. Weird thing all this stuff works correctly when I just use GCC/G++ tools instead of LLVM. So I'm a bit stuck...

Not sure if this helps, but have a look at https://github.com/JCSDA-internal/jedi-bundle/blob/87e5c94134e02a46a3096a6fcafb8707aff2cbe0/CMakeLists.txt#L24 and https://github.com/JCSDA-internal/jedi-bundle/blob/87e5c94134e02a46a3096a6fcafb8707aff2cbe0/CMakeLists.txt#L28 please.

**Comment by rmclaren** [comment]
@climbfuji Did have a look, I'm not using JEDI-bundle (don't need most of the stuff in there) but I did try some of those flags jsn't to see if it makes a difference (doesn't unfortunately). I think the problem is a more fundamental. It has the correct path to the resource in the rpath but fails to recognize it as a valid DYLIB. The library is part of the GCC toolchain (libquadmath.dylib) and I'm sure was compiled with gcc. The problem is that (llvm) clang doesn't like it for some reason... I've compiled the entire toolchain with LLVM on a Ubuntu virtual machine and was able to compile everything with clang there. The problem is that it does not have the problem that Steve witnessed (all the tests pass). So I don't know which regular expression is causing the issue in the Tokenizer...

**Review by BenjaminRuston — APPROVED** [review]
`develop` is passing all ctests for me on my Mac M1 with clang-1400.0.29.202 and  latest stack spack-stack-1.3.0

**Review by BenjaminRuston — COMMENTED** [review]
shooooot,, strike that had accidentally tested `develop` and not `feature/query_filters` am now seeing the same errors as @srherbener

**Comment by emilyhcliu** [comment]
> this is passing all ctests for me on my Mac M1 with clang-1400.0.29.202 and latest stack spack-stack-1.3.0

@rmclaren   Are you seeing the same failure on your Mac?

**Comment by CoryMartin-NOAA** [comment]
@emilyhcliu Ron has an Intel Mac I think, the problems shown here are for Apple Silicon Macs, so the issue may be ARM vs x86_64 architecture related

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (BenjaminRuston) [inline src/bufr/BufrParser/Query/Tokenizer.h:147]: this line does not have  `\\` in the second set of parenthesis different from the Pattern constexpr line above
```suggestion
                std::regex("\\{(\\d+(\\-\\d+)?)(\\,\\d+(\\-\\d+)?)*\\}");
```

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (BenjaminRuston) [inline src/bufr/BufrParser/Query/Tokenizer.h:215]: this one too, I have a regex check that doesn't like the forward slash in from of the right brack

```suggestion
            "([A-Z0-9_\\*\\]+(\\[\\d+\\])?+(\\{[0-9\\-\\,]+\\})?)+";
```

when I change this it does change the error but still failing:

97: libc++abi: terminating with uncaught exception of type std::__1::regex_error: The parser did not consume the entire regular expression.

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (rmclaren) [reply src/bufr/BufrParser/Query/Tokenizer.h:147]: It's actually more correct to remove the `\\` in front of the comma as commas don't need to be escaped (are not treated as special characters in regex).

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (rmclaren) [reply src/bufr/BufrParser/Query/Tokenizer.h:215]: I'd be careful with that regex checker... What tool is it? What variant of regex parser is it checking against? There are unfortunately a lot of regex parsers with different quirks( for example the python regex parser might have different quirks than the PHP regex parser, etc). That forward slash there is pretty damn important :). No way any of this is going to work without that. Also you just escaped the `]` character...

I'll try to sanatize my regex strings a little (make them compatible with more regex parsers). I use tools like https://regex101.com to play around with these. If you try this don't forget to remove the extra `\` characters so for example the string above would become `([A-Z0-9_\*\/]+((\[\d+\])?)+(\{[0-9\-\,]+\})?)+`. You have to escape the escape character in the c++ strings ;).

@BenjaminRuston Are you able to reproduce @srherbener exactly (same error message)?

Comment (rmclaren) [reply src/bufr/BufrParser/Query/Tokenizer.h]: If anyone can reproduce @srherbener's regex problem please try doing a `git pull` and give my latest version a try. I went through and a made my regex strings play better with more types of regex interpreters. I think I might have found one thing that might have caused his error.

**Review by BenjaminRuston — APPROVED** [review]
awesome this now passes all ctest on the M1 mac.

thanks @rmclaren looked at your sanitization and very happy this did the trick!

File: src/bufr/BufrParser/Query/Tokenizer.h
Comment (rmclaren) [reply src/bufr/BufrParser/Query/Tokenizer.h]: @BenjaminRuston Thanks Ben! I spent way to much time trying to recreate the issue on my mahcine... ugg. Thanks for your help :)

**Pull request merged**
